### DISCUSSION TITLE
Building a web interface to manage some IOT devices

### DISCUSSION THREAD

**Discussion by kavitama** [arendst/Tasmota#16695] [Q&A]
Hi,

I am trying to build a web interface using PHP as main language.
I was able to install the php-mqtt-client library and was able to send very basic commands using "tasmota syntax": so far so good.

Even if I am still learning, so don't have a deep knowledge of MQTT, I get that if I want to execute a command I am using a publish command while if I want to know something about the target device I have to subscribe to a topic of my interest.

I am starting with a simple smart plug that has two plugs (power1 and power2).

I get confused about two different items:
- How can I get the acknowledge that a specific publish command has been executed by the device?
- Why I am getting a bad pattern when I try to subscribe?

FIRST
When I send a publish command to switch on power1 I would like to know:
- if the device was receiving the command or was disconnected
- if it receives the command I want a confirmation that the plug is ON
If I understood the pub/sub mechanism I will be "informed" if I am subscribed to the topic that another command is sending as a publish message.
Is this the correct approach or am I missing something big?

SECOND
If the previous assumption is correct, then I am NOT understanding why sending a publish command like this:
`-t cmnd/topic/power1 -m on`
enables the plug one of this smart plug IOT device
but...
when i send the following subscribe command:
`-t cmnd/topic/power1`
I get an error saying I am using a bad patter and the only thing that works is 
`-t '#'`
which shows me everything that the device is sending to the message broker and not simply something like:
`{"POWER1":"ON"}`

I really need help to understand it since the following step will be how to do that using the PHP library

**Comment by barbudor** [comment]
MQTT standard only define messages 1 way and do not imply acknowledgement

The way Tasmota works is that when it receive a command on cmnd/%topic%/COMMAND, it generaly publish back a result on stat/%topic%/RESULT
However with `SetOption4 1`, Tasmota will use the COMMAND in the response instead of RESULT

So for example:
- You publish on topic `cmnd/light1/POWER` the message (payload) `ON`
- Tasmota will reply with on topic `stat/light1/POWER` with mesage (payload) `ON`
Actually for power, Tasmota replies twice : Once with payload `ON` and also with JSON payload `{"POWER":"ON"}` (there is a setoption to disable the text only answer but I don't remember it)

Which means that to receive the answer you should subscribe to at least the topic `stat/light1/POWER`
Now depending on the architecture of your code you can subscribe to various individual topics like `stat/light1/POWER` or use a wildcard like `stat/light1/#` to receive all stat messages from "light1"
You can also subscribe to `stat/+/#`to receive all stat messages from every tamsota

No it depends on your code architecture, if it's multithreaded or not etc ....

Regarding:
> when i send the following subscribe command:
> -t cmnd/iotid/power1
> I get an error saying I am using a bad patter and the only thing that works is

Who gives an error ?
At least the principle is not correct as to toalk to Tasmota you publish on `cmnd/%topic%/COMMAND`
But to listen to Tasmota you need to subscribe to `stat/%topic%/xxxx` and not to 'cmnd/...`

In theory subscribing to `cmnd/%topic%/power1` should not throw an error but would just not provide you with the response from Tasmota but a copy of any command sent to Tasmota by another process.

So as a summary:
To send command to tasmota : publish to `cmnd/%topic%/COMMAND`
To listen to tasmota response : subscribe to `stat/%topic%/COMMAND` (if using `SetOption4 1`) or `stat/%topic%/RESULT` (if not using `Setoption4 1`). You can listen to all responses from tasmota by subscribing to `stat/%topic%/#` for 1 device or `stat/+/#` for all devices
To liste to tasmota telemetry messages subscribe to `tele/%topic%/#` to receive SENSOR and STATE messages

This is the basics for 90% of Tasmota usage (lights, relays, blinds and shutters, switches, sensors, ...)
Bluetooth and Zigbess gateways can be much more complicated 😄

**Comment by kavitama** [comment]
PERFECT!
I also updated the original question using "topic" to be sure I am not confusing someone else about it :-)

At least the first issue is clear to me and it is working when I test it from CLI.

I send a publish command like this:
`-t cmnd/topic/power1 -m on`
which enables the plug one

Using the following subscribe command:
`-t stat/topic/result`
I get the following message:
`{"POWER1":"ON"}`

Which is exactly what i wanted to know

Now more specifically...
When the user presses on a button that says "Switch on Power1" I am using the php-mqtt-client to send the publish message but how I send the subscribe at the same time and get the {"POWER1":"ON"} answer ???
Havent found a decent example in PHP whatsoever.

**Reply by barbudor** [reply]
I've never used php but in python I am subscribing to stat before publishing to cmnd, then check if I'm receiving something on my subscription

It is all the difficulty of an asynchronous protocol
Not has easy as http

**Reply by sfromis** [reply]
You do not "send" a subscribe, but you setup a subscription before doing anything else. You do not get the publish from Tasmota specifically linked to your publish, instead setting up the subscribe means that anything Tasmota is publishing to matching topics while the subscription is active will be picked up. Due to the asynchronous nature of the pubsub model, it is not request/response, but just bidirectional message flows.

**Reply by kavitama** [reply]
> You do not "send" a subscribe, but you setup a subscription before doing anything else. You do not get the publish from Tasmota specifically linked to your publish, instead setting up the subscribe means that anything Tasmota is publishing to matching topics while the subscription is active will be picked up. Due to the asynchronous nature of the pubsub model, it is not request/response, but just bidirectional message flows.

That is exactly the doubt i had but I am not yet there still have to solve how to run two processes from PHP :-(

**Reply by sfromis** [reply]
A good tool for handling MQTT message flows is Node-Red, which has the framework to handle the multithreading without lower level coding. Of course, you still need to decide what message from the other end as a reaction you want to treat as a reaction.
